functions/inputs/from.go

```diff
@@ -178,9 +178,41 @@ func (rule MergeFromRangeRule) Rewrite(node plan.PlanNode) (plan.PlanNode, bool)
 	// Finally merge nodes into single operation
 	merged, err := plan.MergePhysicalPlanNodes(node, from, fromRange)
 	if err != nil {
-		return node, false
+		return nil, false, err
 	}
-	return merged, true
+
+	return merged, true, nil
+}
+
+type FromFilterMergeRule struct {
+}
+
+func (FromFilterMergeRule) Name() string {
+	return "mergeFilterFrom"
+}
+
+func (FromFilterMergeRule) Pattern() plan.Pattern {
+	return plan.Pat(transformations.FilterKind, plan.Pat(FromKind))
+}
+
+func (FromFilterMergeRule) Rewrite(node plan.PlanNode) (plan.PlanNode, bool, error) {
+	filterSpec := node.ProcedureSpec().(*transformations.FilterProcedureSpec)
+	fromSpec := node.Predecessors()[0].ProcedureSpec().(*FromProcedureSpec)
+
+	mergedSpec := fromSpec.Copy().(*FromProcedureSpec)
+	if mergedSpec.FilterSet {
+		mergedSpec.Filter = semantic.MergeArrowFunction(fromSpec.Filter, filterSpec.Fn)
+	} else {
+		mergedSpec.FilterSet = true
+		mergedSpec.Filter = filterSpec.Fn.Copy().(*semantic.FunctionExpression)
+	}
+
+	merged, err := plan.MergePhysicalPlanNodes(node, node.Predecessors()[0], mergedSpec)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return merged, true, nil
```

Comment:
`Rewrite` might need to check if a `range` has been pushed down to `from`. I think storage can support tag filters without a range, but not field filters. Currently we don't know if a predicate is a tag predicate or a field predicate (aside from `_value`). So for right now, we probably can't push down unless a range has already been pushed.

Reply:
I need to explore this a bit more, especially after our recent discussion.